Source/DafnyTestGeneration/Utils.cs

```diff
@@ -1,34 +1,165 @@
-using System;
+#nullable disable
 using System.Collections.Generic;
-using System.Collections.Immutable;
+using System.IO;
 using System.Linq;
-using System.Text;
+using DafnyServer.CounterexampleGeneration;
+using Microsoft.Boogie;
 using Microsoft.Dafny;
 using Errors = Microsoft.Dafny.Errors;
+using Function = Microsoft.Dafny.Function;
 using Parser = Microsoft.Dafny.Parser;
 using Program = Microsoft.Dafny.Program;
+using Token = Microsoft.Dafny.Token;
+using Type = Microsoft.Dafny.Type;
 
 namespace DafnyTestGeneration {
 
   public static class Utils {
 
+    /// <summary>
+    /// Take a resolved type and change all names to fully-qualified.
+    /// </summary>
+    public static Type UseFullName(Type type) {
+      return DafnyModelTypeUtils
+        .ReplaceType(type, _ => true, typ => new UserDefinedType(
+          new Token(),
+          typ?.ResolvedClass?.FullName ?? typ.Name,
+          typ.TypeArgs));
+    }
+
+    /// <summary>
+    /// Copy a <param name="type"></param> and recursively replace type
+    /// arguments named as in <param name="from"></param> with types from
+    /// <param name="to"></param>.
+    /// </summary>
+    public static Type CopyWithReplacements(Type type, List<string> from, List<Type> to) {
+      if (from.Count != to.Count) {
+        return type;
+      }
+      Dictionary<string, Type> replacements = new();
+      for (int i = 0; i < from.Count; i++) {
+        replacements[from[i]] = to[i];
+      }
+      replacements["_System.string"] =
+        new UserDefinedType(new Token(), "string", new List<Type>());
+      replacements["_System.nat"] =
+        new UserDefinedType(new Token(), "nat", new List<Type>());
+      replacements["_System.object"] =
+        new UserDefinedType(new Token(), "object", new List<Type>());
+      return DafnyModelTypeUtils.ReplaceType(type, _ => true,
+        typ => replacements.ContainsKey(typ.Name) ?
+          replacements[typ.Name] :
+          new UserDefinedType(typ.tok, typ.Name, typ.TypeArgs));
+    }
+
     /// <summary>
     /// Parse a string read (from a certain file) to a Dafny Program
     /// </summary>
-    public static Program? Parse(string source, string fileName = "") {
+    public static Program/*?*/ Parse(string source, string fileName = "") {
```

Comment:
Again a step backwards?

Reply:
I removed nullability annotations mostly because they would cause warning-as-errors when interacting with code that does not use them, e.g. one in DafnyLanguageServer. Will see if I can bring them back!